src/v/utils/fragmented_vector.h

```diff
@@ -0,0 +1,133 @@
+ *
+ * There is no reserve method because we allocate a full fragment at a time.
+ * However, after you populate a vector you might want to call shrink to fit if
+ * your fragment is large. A more advanced strategy could allocate capacity up
+ * front which just requires a bit more accounting.
+ *
+ * The iterator implementation works for a few things like std::lower_bound,
+ * upper_bound, distance, etc... see fragmented_vector_test.
+ */
+template<typename T, size_t fragment_size = 8192>
+class fragmented_vector {
+    static_assert(fragment_size % sizeof(T) == 0);
+    static constexpr size_t elems_per_frag = fragment_size / sizeof(T);
+    static_assert(elems_per_frag >= 1);
+
+public:
+    fragmented_vector() noexcept = default;
+    fragmented_vector(const fragmented_vector&) noexcept = delete;
+    fragmented_vector& operator=(const fragmented_vector&) noexcept = delete;
+    fragmented_vector(fragmented_vector&&) noexcept = default;
+    fragmented_vector& operator=(fragmented_vector&&) noexcept = default;
+    ~fragmented_vector() noexcept = default;
+
+    void push_back(T elem) {
+        if (_size == _capacity) {
+            std::vector<T> frag;
+            frag.reserve(elems_per_frag);
+            _frags.push_back(std::move(frag));
+            _capacity += elems_per_frag;
+        }
+        _frags.back().push_back(elem);
+        ++_size;
+    }
+
+    void pop_back() {
+        vassert(_size > 0, "Cannot pop from empty container");
+        _frags.back().pop_back();
+        --_size;
+        if (_frags.back().empty()) {
+            _frags.pop_back();
+            _capacity -= elems_per_frag;
+        }
+    }
+
+    const T& operator[](size_t index) const noexcept {
+        vassert(index < _size, "Index out of range {}/{}", index, _size);
+        auto& frag = _frags.at(index / elems_per_frag);
+        return frag.at(index % elems_per_frag);
+    }
+
+    const T& back() const { return _frags.back().back(); }
+    bool empty() const { return _size == 0; }
+    size_t size() const { return _size; }
+
+    void shrink_to_fit() {
+        if (!_frags.empty()) {
+            _frags.back().shrink_to_fit();
+        }
+    }
+
+    bool operator==(const fragmented_vector& o) const noexcept {
+        return o._frags == _frags;
+    }
+
+    class const_iterator {
+    public:
+        using iterator_category = std::random_access_iterator_tag;
+        using value_type = const T;
+        using difference_type = ssize_t;
+        using pointer = const T*;
+        using reference = const T&;
+
+        reference operator*() const { return _vec->operator[](_index); }
+
+        const_iterator& operator+=(ssize_t n) {
+            _index += n;
+            return *this;
+        }
+
+        const_iterator& operator++() {
+            ++_index;
+            return *this;
+        }
+
+        bool operator==(const const_iterator& o) const {
+            return o._index == _index;
+        }
```

Comment:
nit: can be default, the generated operator will also check that both iterators belong to the same vector
```suggestion
        bool operator==(const const_iterator&) const noexcept = default;
```

Reply:
nice. fixed